source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1027,6 +1028,185 @@ ClusterManagerImpl::addThreadLocalClusterUpdateCallbacks(ClusterUpdateCallbacks&
+ClusterManagerImpl::ClusterDiscoveryManager::addCallback(
+    const std::string& name, ClusterDiscoveryCallbackWeakPtr weak_callback) {
+  auto& callbacks_deque = pending_clusters_[name];
+  callbacks_deque.emplace_back(weak_callback);
+  auto handle =
+      std::make_unique<ClusterDiscoveryCallbackHandleImpl>(*this, std::move(weak_callback), name);
+  auto discovery_in_progress = (callbacks_deque.size() > 1);
+  return {std::move(handle), discovery_in_progress};
+}
+
+void ClusterManagerImpl::ClusterDiscoveryManager::erase(const std::string& name,
+                                                        ClusterDiscoveryCallbackWeakPtr cb) {
+  const bool drop_deque = eraseFromDeque(name, cb);
+  if (drop_deque) {
+    pending_clusters_.erase(name);
+  }
+  maybePostResetCallbacks();
+}
+
+bool ClusterManagerImpl::ClusterDiscoveryManager::eraseFromDeque(
+    const std::string& name, ClusterDiscoveryCallbackWeakPtr cb) {
+  auto it = pending_clusters_.find(name);
+  if (it == pending_clusters_.end()) {
+    return false;
+  }
+  auto& deque = it->second;
+  // Could use std::erase_if, but it's only in C++20.
+  auto it2 = std::remove_if(deque.begin(), deque.end(),
+                            [cb](ClusterDiscoveryCallbackWeakPtr const& weak_ptr) {
+                              if (cb.owner_before(weak_ptr)) {
+                                return false;
+                              }
+                              if (weak_ptr.owner_before(cb)) {
+                                return false;
+                              }
+                              return true;
+                            });
+  deque.erase(it2, deque.end());
+  return deque.empty();
+}
+
+void ClusterManagerImpl::ClusterDiscoveryManager::maybePostResetCallbacks() {
+  if (!callbacks_handle_cleanup_posted_ && pending_clusters_.empty()) {
+    parent_.thread_local_dispatcher_.post([this] {
+      // Something might got added in the meantime, so check the
+      // map again.
+      if (pending_clusters_.empty()) {
+        callbacks_handle_.reset();
+        callbacks_.reset();
+      }
+      callbacks_handle_cleanup_posted_ = false;
+    });
+    callbacks_handle_cleanup_posted_ = true;
+  }
+}
+
+ClusterDiscoveryCallbackHandlePtr
+ClusterManagerImpl::requestOnDemandClusterDiscovery(OdCdsApiSharedPtr odcds,
+                                                    const std::string& name,
+                                                    ClusterDiscoveryCallbackWeakPtr weak_callback) {
+  ThreadLocalClusterManagerImpl& cluster_manager = *tls_;
+
+  cluster_manager.cdm_.ensureCallbacksAreInstalled();
+
+  auto [handle, discovery_in_progress] = cluster_manager.cdm_.addCallback(name, weak_callback);
+  if (discovery_in_progress) {
+    // We have already started a discovery process for a cluster with
+    // this name, so nothing more left to do here.
+    return std::move(handle);
+  }
+  dispatcher_.post([this, odcds = std::move(odcds), weak_callback, name,
```

Comment:
As per my other comment, I think this is tricky cross-thread lifetime related code for on-demand that has _already_ been addressed in both VHDS and SRDS. Re-inventing these patterns here is problematic, we are building guaranteed technical debt.

Sorry to be a downer, I appreciate the work being done here, but I know from experience how hard it was to get things like on-demand VHDS right, we need to have common patterns for deferral and resumption across the on-demand xDS.

CC @mattklein123 @adisuissa @dmitri-d

Reply:
Yes I agree if we can share this out that would be good.

Reply:
I had a look and what VHDS is doing and I can see the differences:

- I'm "deduplicating" the requests, so 2 worker threads asking for the cluster "foo" will result in one request instead of two.
- I'm installing a timeout on the discovery, so the callbacks list won't grow indefinitely in case of cluster "foo" nonexistence.

I guess I can simplify the code, but maybe let's discuss it in the other PR, where I split out the CM changes.